### PR TITLE
fix field removal in mutating admission webhooks

### PR DESCRIPTION
**Pull request by rojkov**
A mutating admission controller webhook doesn't remove object fields
when instructed to.

E.g. when the JSON patch
```
  [
    {"op": "remove", "path": "/spec/containers/0/resources/limits/fpga-arria10"},
    {"op": "add", "path": "/spec/containers/0/resources/limits/fpga-interface-id-524abcf", "value": 1}
  ]
```
is applied to this pod

```
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  restartPolicy: Never
  containers:
    -
      name: test-pod-container
      image: ubuntu:bionic
      imagePullPolicy: IfNotPresent
      command: [ "ls", "-l", "/" ]
      resources:
        limits:
          fpga-arria10: 1
```

in order to replace the resource name "fpga-arria10" with something understandable
by the device plugin the resulting pod spec still contains the old field plus
a new one. The resulting pod looks like
```
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  restartPolicy: Never
  containers:
    -
      name: test-pod-container
      image: ubuntu:bionic
      imagePullPolicy: IfNotPresent
      command: [ "ls", "-l", "/" ]
      resources:
        limits:
          fpga-arria10: 1
          fpga-interface-id-524abcf: 1
```

The patch unmarshals patched JSON into a new empty object instead of
existing one. Otherwise JSON unmarshaling reuses existing maps, keeping
existing entries as specified in the "encoding/json" standard package.

**Release note**:
```release-note
Webhooks for the mutating admission controller now support "remove" operation.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 1]
```diff
@@ -113,11 +116,22 @@ func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta
...
+			return apierrors.NewInternalError(err)
+		}
+	}
 	// TODO: if we have multiple mutating webhooks, we can remember the json
 	// instead of encoding and decoding for each one.
-	if _, _, err := a.plugin.jsonSerializer.Decode(patchedJS, nil, attr.VersionedObject); err != nil {
+	if _, _, err := a.plugin.jsonSerializer.Decode(patchedJS, nil, newVersionedObject); err != nil {
 		return apierrors.NewInternalError(err)
 	}
+	attr.VersionedObject = newVersionedObject
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: can you assign directly to `attr.VersionedObject` above?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin_test.go [hunk 2]
```diff
@@ -46,6 +49,8 @@ func TestAdmit(t *testing.T) {
 	defer close(stopCh)
 
 	for _, tt := range webhooktesting.NewTestCases(serverURL) {
+		var attr admission.Attributes
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/plugin_test.go]: move this down where it is assigned

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go [hunk 3]
```diff
@@ -106,6 +113,35 @@ func NewAttribute(namespace string) admission.Attributes {
 	return admission.NewAttributesRecord(&object, &oldObject, kind, namespace, name, resource, subResource, operation, &userInfo)
 }
 
+// NewAttributeUnstructured returns static admission Attributes for testing with custom resources.
+func NewAttributeUnstructured(namespace string) admission.Attributes {
+	object := unstructured.Unstructured{}
+	name := "my-test-crd"
+	object.SetKind("TestCRD")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go]: can we unify both functions by using a metadata accessor (works for both)?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go]: Also it is odd that this one does not take labels.

**Comment by sttts** [comment]
Looks very reasonable. Left some comments about style and uniformity.

**Comment by sttts** [comment]
This will also need a backport to 1.10.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: I actually prefer waiting to overwrite attr.VersionedObject until the decode succeeds

don't we still have the problem that converting attr.VersionedObject back into `attr.Attributes.GetObject()` doesn't necessarily clear the original object?

https://github.com/kubernetes/kubernetes/blob/a7a081b5410fcf711b1f85169225974040051204/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go#L75-L76

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: > do we still have a problem when ...

You mean "a similar problem in addition", right?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go [hunk 4]
```diff
@@ -78,6 +79,8 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr *generic.Version
 
 // note that callAttrMutatingHook updates attr
 func (a *mutatingDispatcher) callAttrMutatingHook(ctx context.Context, h *v1beta1.Webhook, attr *generic.VersionedAttributes) error {
+	var newVersionedObject runtime.Object
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: please move this down where it is used.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: > You mean "a similar problem in addition", right?

yes

**Comment by lavalamp** [comment]
We should cherry pick this. It could cause a lot of problems. Thank you for finding it!

I think the release note could be expanded--I think this should cause more problems than just inability to remove a field? And it only shows up if there's more than one mutating webhook, correct?

**Comment by rojkov** [comment]
@lavalamp I'm enabling various hardware features in K8s at Intel and don't run complex installations :) I stumbled upon the issue in my devel setup with only one mutating webhook.

I'll address the review comments on Monday.

**Comment by jberkus** [comment]
Hey, do you plan to merge this to 1.11?  currently it needs approvals, milestone, and priority.

**Comment by lavalamp** [comment]
We will cherry-pick if it doesn't make it.

**Comment by sttts** [comment]
@wgliang are you able to address comments early next week? We would like to get this into 1.11.

**Comment by wgliang** [comment]
@rojkov This is assign to you. If you need help, please reply here.

> _are you able to address comments early next week? We would like to get this into 1.11_

 from @sttts

**Comment by rojkov** [comment]
1. moved variable declarations closer to their first use;
2. added more test cases covering addition and deletion of lables in CRDs to justify the `labels` patameter of `NewAttributeUnstructured()`;
3. simplified the code testing equivalence of object lables in TestAdmin() (previously the code was different for CRD and non-CRD objects);
4. unified `NewAttribute()` and `NewAttributeUnstructured()` to use metadata accessors for setting up test objects.
5. rebased to the latest master

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go (hunk 1 above)
Comment (rojkov) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: I would leave this `newVersionedObject` as it is too, but my rationale is different. With the brand new empty variable it's clearer that the call `a.plugin.jsonSerializer.Decode()` is supposed to update the variable and not to update the internal state of the called object. For me it took a while to understand where exactly attr.VersionedObject gets updated when I saw the code for the first time.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go]: > do we still have a problem when converting attr.VersionedObject back into attr.Attributes.GetObject() if conversion doesn't necessarily clear/reset the original internal object?

This needs a follow-up investigation.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go [hunk 5]
```diff
@@ -74,36 +76,65 @@ func NewFakeDataSource(name string, webhooks []registrationv1beta1.Webhook, muta
 	return client, informerFactory
 }
 
+func newAttributesRecord(object metav1.Object, oldObject metav1.Object, kind schema.GroupVersionKind, namespace string, name string, resource string, labels map[string]string) admission.Attributes {
+	object.SetName(name)
+	object.SetNamespace(namespace)
+	objectLabels := map[string]string{resource + ".name": name}
+	if labels != nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go]: nit: no need for the check

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go [hunk 6]
```diff
@@ -74,36 +76,65 @@ func NewFakeDataSource(name string, webhooks []registrationv1beta1.Webhook, muta
...
+		}
+	}
+	object.SetLabels(objectLabels)
+
+	oldObject.SetName(name)
+	oldObject.SetNamespace(namespace)
+
+	operation := admission.Update
+
+	gvr := schema.GroupVersionResource{
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go]: isn't there `kind.GroupVersion().WithResource(resource)`?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go [hunk 7]
```diff
@@ -74,36 +76,65 @@ func NewFakeDataSource(name string, webhooks []registrationv1beta1.Webhook, muta
...
+		for k, v := range labels {
+			objectLabels[k] = v
+		}
+	}
+	object.SetLabels(objectLabels)
+
+	oldObject.SetName(name)
+	oldObject.SetNamespace(namespace)
+
+	operation := admission.Update
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go]: nit: put it directly into the return line.

**Comment by dims** [comment]
/priority critical-urgent

(per note from @sttts on sig-release)

"it’s fixing a serious bug in admission webhooks. if we don’t get it into 1.11.0 ppl have to work around it in a 1.11.0 somehow which is ugly. and if ppl rely on correct behaviour, it turns into a security issue."

**Comment by sttts** [comment]
priority/important-soon

**Pull request merged**